tests/test_modeling_tf_common.py

```diff
@@ -380,65 +366,191 @@ def test_pt_tf_model_equivalence(self):
                 else:
                     pt_inputs_dict[name] = torch.from_numpy(key.numpy()).to(torch.long)
 
-            with torch.no_grad():
-                pto = pt_model(**pt_inputs_dict)
-            tfo = tf_model(self._prepare_for_class(inputs_dict, model_class), training=False)
+            return pt_inputs_dict
+
+        def check_outputs(tfo, pto, model_class, names):
+            """
+            Args:
+                model_class: The class of the model that is currently testing. For example, `TFBertModel`,
+                    TFBertForMaskedLM`, `TFBertForSequenceClassification`, etc. Currently unused, but it could make
+                    debugging easier and faster.
+
+                names: A string, or a list of strings. These specify what tfo/pto represent in the model outputs.
+                    Currently unused, but in the future, we could use this information to make the error message clearer
+                    by giving the name(s) of the output tensor(s) with large difference(s) between PT and TF.
+            """
+
+            # Some issue (`about past_key_values`) to solve (e.g. `TFPegasusForConditionalGeneration`) in a separate PR.
+            if names == "past_key_values":
+                return
+
+            if type(tfo) == tuple:
+                self.assertEqual(type(pto), tuple)
+                self.assertEqual(len(tfo), len(pto))
+                if type(names) in [tuple, list]:
+                    for to, po, name in zip(tfo, pto, names):
+                        check_outputs(to, po, model_class, names=name)
+                elif type(names) == str:
+                    for idx, (to, po) in enumerate(zip(tfo, pto)):
+                        check_outputs(to, po, model_class, names=f"{names}_{idx}")
+            elif isinstance(tfo, tf.Tensor):
+                self.assertTrue(isinstance(pto, torch.Tensor))
+
+                tfo = tfo.numpy()
+                pto = pto.numpy()
+
+                tf_nans = np.copy(np.isnan(tfo))
+                pt_nans = np.copy(np.isnan(pto))
+
+                pto[tf_nans] = 0
+                tfo[tf_nans] = 0
+                pto[pt_nans] = 0
+                tfo[pt_nans] = 0
+
+                max_diff = np.amax(np.abs(tfo - pto))
+                self.assertLessEqual(max_diff, 1e-5)
```

Comment:
Will we need some kind of relative tolerance here? 1e-5 is a small allowable difference for potentially large values!

Reply:
Probably yes. Currently I have confidence that `1e-5` would work in our current testing configs/models.
The weights are initialized with a std 0.02 -> and with this setting, the output values won't get too large.
(Lysandre told me that once we go for GPU testing and fp16 precision testing, we might need to deal with larger errors).

Reply:
Yes, would love to know if this passes on GPU as it's typically tougher on small differences.

The point regarding fp16 was specifically regarding bf16: models trained with bf16 typically have much larger logits, so using an absolute difference is unideal; using relative difference should be used.

Here we do the init ourselves, so it doesn't apply.